EventFilter/Phase2TrackerRawToDigi/plugins/DTHDAQToFEDRawDataConverter.cc

```diff
@@ -0,0 +1,151 @@
+#include "FWCore/Framework/interface/one/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "DataFormats/FEDRawData/interface/FEDRawData.h"
+#include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include <fstream>
+#include <vector>
+#include <iomanip>
+#include <cstdint>
+
+
+// Constants for bit field widths, markers, and sizes
+constexpr unsigned int orbitCount = 4;
+constexpr unsigned int orbitHeaderSize = 32;
+constexpr unsigned int fragmentTrailerSize = 16;
+constexpr unsigned int fragmentPayloadWordSize = 16;
+constexpr unsigned int orbitVersionSize = 2;
+constexpr unsigned int sourceIdSize = 4;
+constexpr unsigned int runNumberSize = 4;
+constexpr unsigned int orbitNumberSize = 4;
+constexpr unsigned int eventCountResSize = 4;
+constexpr unsigned int packetWordCountSize = 4;
+constexpr unsigned int flagsSize = 4;
+constexpr unsigned int checksumSize = 4;
+constexpr unsigned int fragFlagSize = 2;
+constexpr unsigned int fragSizeSize = 4;
+constexpr uint8_t orbitHeaderMarkerH = 0x48;
+constexpr uint8_t orbitHeaderMarkerO = 0x4F;
+constexpr uint8_t fragmentTrailerMarkerF = 0x48;
+constexpr uint8_t fragmentTrailerMarkerH = 0x46;
+
+class DTHDAQToFEDRawDataConverter : public edm::one::EDProducer<> {
+public:
+  explicit DTHDAQToFEDRawDataConverter(const edm::ParameterSet&);
+  ~DTHDAQToFEDRawDataConverter() override = default;
+
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+private:
+  std::string inputFile_;
+  unsigned int fedId_;
+
+  std::vector<char> readRawFile(const std::string& inputFile);
+  void parseAndDumpEventData(const std::vector<char>& buffer);
+  void reverseParseFragments(const std::vector<char>& buffer, size_t startIdx, uint32_t packetWordCount, uint16_t eventCount);
+  void printHex(const std::vector<char>& buffer, size_t length);
+};
+
+DTHDAQToFEDRawDataConverter::DTHDAQToFEDRawDataConverter(const edm::ParameterSet& config)
+  : inputFile_(config.getParameter<std::string>("inputFile")),
+    fedId_(config.getParameter<unsigned int>("fedId")) {
+  produces<FEDRawDataCollection>();
+}
+
+std::vector<char> DTHDAQToFEDRawDataConverter::readRawFile(const std::string& inputFile) {
+  std::ifstream rawFile(inputFile, std::ios::binary | std::ios::ate);
+  if (!rawFile.is_open()) {
+    throw cms::Exception("FileOpenError") << "Could not open input file: " << inputFile;
+  }
+
+  std::streamsize fileSize = rawFile.tellg();
+  rawFile.seekg(0, std::ios::beg);
+
+  std::vector<char> buffer(fileSize);
```

Comment:
You seem to be reading the whole file, which may consist of many orbits into a buffer. This will take huge amounts of memory. Why not only read the minimum data you need in order to extract the first event. (e.g. Perhaps data from the first orbit?)